internal/provider/resource_tfe_registry_module.go

```diff
@@ -416,15 +417,30 @@ func resourceTFERegistryModuleRead(d *schema.ResourceData, meta interface{}) err
 func resourceTFERegistryModuleDelete(d *schema.ResourceData, meta interface{}) error {
 	config := meta.(ConfiguredClient)
 
-	log.Printf("[DEBUG] Delete registry module: %s", d.Id())
-	organization := d.Get("organization").(string)
-	name := d.Get("name").(string)
-	err := config.Client.RegistryModules.Delete(ctx, organization, name)
-	if err != nil {
-		if err == tfe.ErrResourceNotFound {
-			return nil
+	// Fields required to delete registry module by provider
+	// To delete by name, Provider field is not required
+	rModID := tfe.RegistryModuleID{
+		Organization: d.Get("organization").(string),
+		Name:         d.Get("name").(string),
+		Provider:     d.Get("module_provider").(string),
+		Namespace:    d.Get("namespace").(string),
+		RegistryName: tfe.RegistryName(d.Get("registry_name").(string)),
+	}
+
+	if v, ok := d.GetOk("module_provider"); ok && v.(string) != "" {
+		log.Printf("[DEBUG] Delete registry module by provider: %s", d.Id())
+
+		err := config.Client.RegistryModules.DeleteProvider(ctx, rModID)
+		if err != nil && !errors.Is(err, tfe.ErrResourceNotFound) {
+			return fmt.Errorf("error deleting registry module provider: %w", err)
+		}
+	} else {
+		log.Printf("[DEBUG] Delete registry module by name: %s", d.Id())
+
+		err := config.Client.RegistryModules.DeleteByName(ctx, rModID)
+		if err != nil && !errors.Is(err, tfe.ErrResourceNotFound) {
+			return fmt.Errorf("Error deleting registry module %s: %w", d.Id(), err)
```

Comment:
So I have some pretty strong opinions against mix & matching different endpoints within a single resource operation depending on the attributes defined. While it's totally possible to move forward with this, I would prefer not to introduce the footgun of having either `module_provider` defined resulting in "the specified provider for the given module is deleted along with all its versions" ([docs](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/private-registry/modules#parameters-2)). Or, omitting the `module_provider` attribute which results in the entire module being deleted. 

With any specific Terraform resource, I expect providers to manage a single, real "thing"... be it a relationship between two API resources or the API resource itself. I'd prefer to avoid introducing side effects that can cause undesirable behavior. 

In my opinion, a better approach would be to add new `tfe_registry_module_version`, which manages a specific version for a module, and `tfe_registry_module_provider`, which manages a specific provider for a module. We can then leave this existing resource `tfe_registry_module` to manage the entire module. Each resource would then call one of the three possible delete endpoints for modules. 

cc @brandonc

Reply:
FWIW, the original intent of this PR was to fix a bug where the `Delete()` method was calling a deprecated endpoint (see [docs](https://developer.hashicorp.com/terraform/enterprise/api-docs/private-registry/modules#delete-a-module)). I think for the purpose of resolving that bug for the customer, we can address it by updating the Delete() method to call `DeleteByName` - and then separate any additional work into new tickets.

Reply:
We had an offline conversation and decided that, while it would be better to have discreet records, this is still a valuable bug fix